src/cryptographic_primitives/proofs/sigma_valid_pedersen.rs

```diff
@@ -0,0 +1,128 @@
+
+use cryptographic_primitives::hashing::hash_sha256::HSha256;
+use cryptographic_primitives::hashing::traits::Hash;
+
+use cryptographic_primitives::commitments::pedersen_commitment::PedersenCommitment;
+use cryptographic_primitives::commitments::traits::Commitment;
+
+use elliptic::curves::secp256_k1::Secp256k1Point;
+use elliptic::curves::secp256_k1::Secp256k1Scalar;
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct PedersenProof {
+    e: Secp256k1Scalar,
+    a1: Secp256k1Point,
+    a2: Secp256k1Point,
+    pub com: Secp256k1Point,
+    z1: Secp256k1Scalar,
+    z2: Secp256k1Scalar,
+}
+pub trait ProvePederesen {
+    fn prove(m: &Secp256k1Scalar, r: &Secp256k1Scalar) -> PedersenProof;
+
+    fn verify(proof: &PedersenProof) -> Result<(), ProofError>;
+}
+
+impl ProvePederesen for PedersenProof {
+    fn prove(m: &Secp256k1Scalar, r: &Secp256k1Scalar) -> PedersenProof {
+        let g: Secp256k1Point = ECPoint::new();
+        let h = Secp256k1Point::base_point2();
+        let s1: Secp256k1Scalar = ECScalar::new_random();
+        let s2: Secp256k1Scalar = ECScalar::new_random();
+        let a1 = g.scalar_mul(&s1.get_element());
+        let a2 = h.scalar_mul(&s2.get_element());
+        let com = PedersenCommitment::create_commitment_with_user_defined_randomness(
+            &m.to_big_int(),
+            &r.to_big_int(),
+        );
+        let g: Secp256k1Point = ECPoint::new();
+        let challenge = HSha256::create_hash(vec![
+            &g.get_x_coor_as_big_int(),
+            &Secp256k1Point::base_point2().get_x_coor_as_big_int(),
+            &com.get_x_coor_as_big_int(),
+            &a1.get_x_coor_as_big_int(),
+            &a2.get_x_coor_as_big_int(),
+        ]);
+        let e: Secp256k1Scalar = ECScalar::from_big_int(&challenge);
+        let em = e.mul(&m.get_element());
+        let z1 = s1.add(&em.get_element());
+        let er = e.mul(&r.get_element());
+        let z2 = s2.add(&er.get_element());
+        PedersenProof {
+            e,
+            a1: a1,
+            a2: a2,
+            com,
+            z1,
+            z2,
+        }
+    }
+
+    fn verify(proof: &PedersenProof) -> Result<(), ProofError> {
+        let g: Secp256k1Point = ECPoint::new();
+        let h = Secp256k1Point::base_point2();
+        let challenge = HSha256::create_hash(vec![
+            &g.get_x_coor_as_big_int(),
+            &h.get_x_coor_as_big_int(),
+            &proof.com.get_x_coor_as_big_int(),
+            &proof.a1.get_x_coor_as_big_int(),
+            &proof.a2.get_x_coor_as_big_int(),
+        ]);
+        let e: Secp256k1Scalar = ECScalar::from_big_int(&challenge);
+        let z1g = g.scalar_mul(&proof.z1.get_element());
+        let z2h = h.scalar_mul(&proof.z2.get_element());
+        let lhs = z1g.add_point(&z2h.get_element());
+        let rhs = proof.a1.add_point(&proof.a2.get_element());
+        let com_clone = proof.com.clone();
+        let ecom = com_clone.scalar_mul(&e.get_element());
```

Comment:
there is something slightly unpleasant about `scalar_mul` taking `mut self` while other operations don't. perhaps a `Borrow` could work and still get the same reuse optimisation when needed? in any case the extra clones could be move into that.

Reply:
I'm not sure I follow. scalar_mul takes `self` with the intention to consume `self`. 
This was done to limit the interface. Is this a problem?

Reply:
not a problem, only saying that the clone could be hidden if desired

Reply:
can you give an example please?

Reply:
"hidden" as in moved into the function by letting the function take eg a Cow instead. does that make sense?